### PR TITLE
proto: use shared bytes for coprocessor response.

### PR DESCRIPTION
**Pull request by coocood**
Reduce memory allocation.

### CODE REVIEW

**Comment by shenli** [comment]
Any benchmark result?

**Comment by coocood** [comment]
Same 1 million rows query:

Old:
```
# Alloc = 1617988216
# TotalAlloc = 5759754344
```

After `tipb.Chunk` use shared bytes:
```
# Alloc = 1305848152
# TotalAlloc = 4735086792
```

After `coprocessor.Response` use shared bytes:
```
# Alloc = 209450744
# TotalAlloc = 3710726160
```

**Comment by zimulala** [comment]
Fix CI

File: pkg/util/codec.go [hunk 1]
```diff
@@ -70,3 +59,22 @@ func ReadMessage(r io.Reader, msg proto.Message) (uint64, error) {
...
+	_, err = io.ReadFull(r, header[:])
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+
+	if magic := binary.BigEndian.Uint16(header[0:2]); magic != msgMagic {
+		return 0, 0, errors.Errorf("mismatch header magic %x != %x", magic, msgMagic)
+	}
+
+	// skip version now.
```
Comment (zimulala) [inline pkg/util/codec.go]: s/skip/Skip

File: pkg/util/codec.go [hunk 2]
```diff
@@ -70,3 +59,22 @@ func ReadMessage(r io.Reader, msg proto.Message) (uint64, error) {
 
 	return msgID, nil
 }
+
+// ReadHeader reads a protocal header from reader.
+func ReadHeader(r io.Reader) (msgID uint64, msgLen uint32, err error) {
+	var header [msgHeaderSize]byte
+	_, err = io.ReadFull(r, header[:])
```
Comment (zimulala) [inline pkg/util/codec.go]: s/header[:]/header ?

**Comment by coocood** [comment]
@zimulala 
The CI never PASS because there is no test file.

File: pkg/util/codec.go (hunk 2 above)
Comment (coocood) [inline pkg/util/codec.go]: No, header is an array, header[:] is a slice,

File: pkg/util/codec.go [hunk 3]
```diff
@@ -70,3 +59,22 @@ func ReadMessage(r io.Reader, msg proto.Message) (uint64, error) {
...
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+
+	if magic := binary.BigEndian.Uint16(header[0:2]); magic != msgMagic {
+		return 0, 0, errors.Errorf("mismatch header magic %x != %x", magic, msgMagic)
+	}
+
+	// Skip version now.
```
Comment (ngaut) [inline pkg/util/codec.go]: Could you add a todo?

**Comment by ngaut** [comment]
LGTM， please fix CI.

File: pkg/util/codec.go [hunk 4]
```diff
@@ -70,3 +59,23 @@ func ReadMessage(r io.Reader, msg proto.Message) (uint64, error) {
 
 	return msgID, nil
 }
+
+// ReadHeader reads a protocal header from reader.
```
Comment (shenli) [inline pkg/util/codec.go]: protocal ->protocol

**Comment by zimulala** [comment]
LGTM

**Pull request merged**